crud/delete.lua

```diff
@@ -21,14 +22,57 @@ local function delete_on_storage(space_name, key)
         return nil, DeleteError:new("Space %q doesn't exist", space_name)
     end
 
-    local tuple = space:delete(key)
-    return tuple
+    -- add_space_schema_hash is false because
+    -- reloading space format on router can't avoid delete error on storage
+    return schema.wrap_box_space_func_result(false, space, 'delete', key)
 end
 
 function delete.init()
    _G._crud.delete_on_storage = delete_on_storage
 end
 
+-- returns result, err, need_reload
+-- need_reload indicates if reloading schema could help
+-- see crud.common.schema.wrap_func_reload()
+local function call_delete(space_name, key, opts)
+    dev_checks('string', '?', {
+        timeout = '?number',
+        bucket_id = '?number|cdata',
+    })
+
+    opts = opts or {}
+
+    local space = utils.get_space(space_name, vshard.router.routeall())
+    if space == nil then
+        return nil, DeleteError:new("Space %q doesn't exist", space_name), true
+    end
+
+    if box.tuple.is(key) then
+        key = key:totable()
+    end
+
+    local bucket_id = sharding.key_get_bucket_id(key, opts.bucket_id)
+    local storage_result, err = call.rw_single(
+        bucket_id, DELETE_FUNC_NAME,
+        {space_name, key}, {timeout = opts.timeout}
+    )
+
+    if err ~= nil then
+        return nil, DeleteError:new("Failed to call delete on storage-side: %s", err)
+    end
+
+    if storage_result.err ~= nil then
+        return nil, DeleteError:new("Failed to delete: %s", storage_result.err)
+    end
+
+    local tuple = storage_result.res
+
+    return {
+        metadata = table.copy(space:format()),
+        rows = {tuple},
+    }
```

Comment:
Finally, I don't understand why did you move "format_result" into separate function if you don't actually use it.

Reply:
Oh shit, seems I've missed this changes =(